enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,637 @@
+
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+The CI operator will run an End-to-End test suite against the CCM and also push the resulting image into the OpenShift Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy the OpenStack CCM and test how it works.
+
+Actions:
+
+- Use development steps to install CCM on a working OpenShift cluster deployed on OpenStack with in-tree plugin.
+- Check the CCM functionality is operational at this point.
+
+**Note:** Example of a manual testing: https://asciinema.org/a/303399?speed=2
+
+#### Test Cloud Controller Manager in CI
+
+Automate OpenStack CCM testing in CI. Introduce a number of CI jobs for the provider.
+
+Actions:
+
+- Unit, local OpenStack CCM build.
+- Release image build.
+- Run cluster installation with out-of-tree OpenStack provider by default.
+- Cluster upgrade from OpenStack in-tree to out-of-tree.
+- Run existing e2e tests for the provider code, to confirm it's functionality.
+
+#### Integrate the solution with OpenShift (Step 1)
+
+Actions:
+
+- `CCCMO` supports deploying OpenStack `CCM`.
+- `KCM` runs only in-tree OpenStack storage implmentation.
+
+#### Integrate the solution with OpenShift (Step 2)
+
+Actions:
+
+- Make sure the Cinder CSI driver is supported in OpenShift.
+- Make sure the Cinder CSI migration is supported in OpenShift.
+- `KCM` excludes the rest of cloud-controller loops by running CSI storage out-of-tree.
+
+#### Examples for other cloud-controller-manager configurations
+
+- AWS: https://github.com/kubernetes/cloud-provider-aws/. **Demo:** , to reproduce see: https://github.com/cloud-team-poc/cloud-provider-aws. [![asciicast](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m.svg)](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m)
+- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
+Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
+- GCP: https://github.com/kubernetes/cloud-provider-gcp
+- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
+vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64
+
+### Risks and Mitigations
+
+#### Specific CCM doesn’t work properly on OpenShift
+
+The `CCMs` have not been tested on OCP yet. Unknown issues may arise.
+
+Severity: medium-high
+
+Likelihood: medium
+
+The likelihood of the risk above varies per provider. None of the providers will be shipped without testing.
+
+### Test Plan
+
+**All following test plan details have a hard requirement - support for teach preview FeatureGate described in this [section](#Pre-release-/-Development-Feature-Gate). No development or testing could be done before the functionality is implemented.**
```

Comment:
Adding comment here but it applies throughout, for cloud providers that were never in-tree (e.g. IBM cloud provider), can we have a streamline set of requirements that go directly to the CCM implementation only?